packages/type-utils/src/getAwaitedType.ts

```diff
@@ -0,0 +1,14 @@
+import * as tsutils from 'ts-api-utils';
+import type * as ts from 'typescript';
+
+import { isPromiseLike } from './builtinSymbolLikes';
+
+export function getAwaitedType(program: ts.Program, type: ts.Type): ts.Type {
+  if (isPromiseLike(program, type) && tsutils.isTypeReference(type)) {
```

Comment:
I think that using `isPromiseLike` here is a little too permissive, the rest of this code specifically assumes the `PromiseLike` _is_ a `Promise` (by checking its type argument position).

We can trick the rule into false positives with code like 
```ts
class SneakyPromiseLike<T = any> extends Promise<number> { }

// returns Promise<number>
async function foo() {
  return new SneakyPromiseLike((resolve) => { resolve(42); });
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Unsafe return of an `Promise<any>` typed value. 
}
```

Here's my thought... if we can, let's _only_ worry about `getAwaitedType` working on proper `Promises` rather than `PromiseLike`s, and err on the side of false negative compared to false positive. 

(for example, i think it's ok if this code doesn't flag, even though it would in a perfect world)
```ts
class SneakyPromiseLike extends Promise<any> { }

async function foo() {
  return new SneakyPromiseLike((resolve) => { resolve('any!'); });
}
```

Since my first review of this PR, I've learned that TS's internal `checker.getAwaitedType` method is in the pipeline to become available to us to use in our rules, at which point, we can replace this implementation with the one that best represents TS's semantics. Therefore, I think it's best for this to just be a best-effort for now, rather than completely comprehensive.

What do you think?

Reply:
@kirkwaiblinger Agree 
 This seems to be similar to the [issue](https://github.com/typescript-eslint/typescript-eslint/issues/8433.), I'll fix it.

Reply:
@kirkwaiblinger I agree. I think we can check if it's a promise in the following way.

```ts
  tsutils.isThenableType(checker, node, type) &&
        tsutils.isTypeReference(type)
```

Reply:
My thought is it makes sense to check for exact promises, but i'm saying that with low confidence, since I'm not personally very familiar with the APIs you've mentioned.